### PR TITLE
[5.7] Fix support for guest user gates when defining class@method style callbacks

### PR DESCRIPTION
**Pull request by jlaswell**
Re-submit of https://github.com/laravel/framework/pull/26904.

### CODE REVIEW

**Comment by mfn** [comment]
> Fix support

FTR, IMHO the title isn't correct. There was nothing broken, it simply wans't part of the initial implementation. This "Adds support" for this new feature; nothing is broken to be fixed. #SmallSemantics ;-) You even say so in your first commit message.

**Comment by taylorotwell** [comment]
Why not just use policies in the normal way?

**Comment by jlaswell** [comment]
Guest users will always be denied authorization when a policy is defined using the `class@method` syntax without this PR; here is the initial replicated issue for reference: https://github.com/laravel/framework/issues/26221. Since the docs show this available, this feature does not work as expected.

I would say that this is a fairly normal use of this with it having been illustrated in the docs, but I am happy to make mention of the `class@method` syntax not supporting guest users if you would prefer.

From the docs on [writing gates](https://laravel.com/docs/5.7/authorization#gates).
> Gates may also be defined using a Class@method style callback string, like controllers:
```php
/**
 * Register any authentication / authorization services.
 *
 * @return void
 */
public function boot()
{
    $this->registerPolicies();

    Gate::define('update-post', 'App\Policies\PostPolicy@update');
}
```

From the docs on [guest users](https://laravel.com/docs/5.7/authorization#guest-users).
> ...you may allow these authorization checks to pass through to your gates and policies by declaring an "optional" type-hint or supplying a null default value for the user argument definition:
```php
class PostPolicy
{
    /**
     * Determine if the given post can be updated by the user.
     *
     * @param  \App\User  $user
     * @param  \App\Post  $post
     * @return bool
     */
    public function update(?User $user, Post $post)
    {
        return $user->id === $post->user_id;
    }
}
```

**Comment by IsraelOrtuno** [comment]
As @jlaswell says, the docs suggested I could use it that way and then I found the "bug" #26221

File: src/Illuminate/Auth/Access/Gate.php [hunk 1]
```diff
@@ -113,6 +120,9 @@ public function define($ability, $callback)
             $this->abilities[$ability] = $callback;
         } elseif (is_string($callback)) {
             $this->abilities[$ability] = $this->buildAbilityCallback($ability, $callback);
+            if ($this->abilities[$ability] instanceof \Closure) {
```
Comment (taylorotwell) [inline src/Illuminate/Auth/Access/Gate.php]: Why do you need to check for this here? Doesn't `buildAbilityCallback` always return a Closure? I'm still a little unsure why you need to maintain this separate list of Closure callbacks.

**Comment by taylorotwell** [comment]
Closing pending inactivity. Feel free to re-open.

**Pull request closed without merging**